File: backend/app/schemas/appeal.py
"""Pydantic schemas cho Phúc khảo kết quả (Result Appeal) — 2026-06-27."""

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

from app.models.enums import AppealStatus


class AppealCreateIn(BaseModel):
    """SV gửi phúc khảo cho 1 entry của mình."""

    entry_id: int
    round_id: int | None = None
    title: str = Field(..., min_length=3, max_length=255)
    content_text: str = Field(..., min_length=5)


class AppealResolveIn(BaseModel):
    """GV/BTC kết luận phúc khảo.

    decision=ACCEPTED  → chấp nhận, kết quả sẽ được chấm lại + duyệt lại + publish lại.
    decision=REJECTED  → giữ nguyên kết quả (kể cả khi đã xem lại thấy điểm đúng).
    response_text bắt buộc để SV biết lý do.
    """

    decision: Literal["ACCEPTED", "REJECTED"]
    response_text: str = Field(..., min_length=5)


class AppealWindowIn(BaseModel):
    """BTC đặt/đổi hạn nhận phúc khảo cho contest. None = đóng kênh phúc khảo."""

    appeal_deadline: datetime | None = None


class AppealOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    appeal_id: int
    contest_id: int
    round_id: int | None = None
    entry_id: int
    submitted_by_student_id: int
    title: str
    content_text: str
    status: AppealStatus
    response_text: str | None = None
    handled_by: int | None = None
    handled_at: datetime | None = None
    created_at: datetime
    updated_at: datetime

File: frontend/lib/core/models/appeal.dart
/// Model phúc khảo kết quả (Result Appeal) — 2026-06-27.
class AppealModel {
  final int appealId;
  final int contestId;
  final int? roundId;
  final int entryId;
  final int submittedByStudentId;
  final String title;
  final String contentText;
  final String status; // PENDING / IN_REVIEW / ACCEPTED / REJECTED / CLOSED
  final String? responseText;
  final int? handledBy;
  final DateTime? handledAt;
  final DateTime createdAt;
  final DateTime updatedAt;

  AppealModel({
    required this.appealId,
    required this.contestId,
    this.roundId,
    required this.entryId,
    required this.submittedByStudentId,
    required this.title,
    required this.contentText,
    required this.status,
    this.responseText,
    this.handledBy,
    this.handledAt,
    required this.createdAt,
    required this.updatedAt,
  });

  factory AppealModel.fromJson(Map<String, dynamic> j) => AppealModel(
        appealId: j['appeal_id'],
        contestId: j['contest_id'],
        roundId: j['round_id'],
        entryId: j['entry_id'],
        submittedByStudentId: j['submitted_by_student_id'],
        title: j['title'] ?? '',
        contentText: j['content_text'] ?? '',
        status: j['status'] ?? 'PENDING',
        responseText: j['response_text'] as String?,
        handledBy: j['handled_by'],
        handledAt: j['handled_at'] != null
            ? DateTime.parse(j['handled_at'])
            : null,
        createdAt: DateTime.parse(j['created_at']),
        updatedAt: DateTime.parse(j['updated_at']),
      );

  /// Nhãn tiếng Việt cho trạng thái.
  String get statusLabel => switch (status) {
        'PENDING' => 'Chờ xử lý',
        'IN_REVIEW' => 'Đang xem xét',
        'ACCEPTED' => 'Được chấp nhận',
        'REJECTED' => 'Bị từ chối',
        'CLOSED' => 'Đã đóng',
        _ => status,
      };

  bool get isOpen => status == 'PENDING' || status == 'IN_REVIEW';
}
